File: src/main/java/no/nav/sbl/dialogarena/mininnboks/provider/rest/ubehandletmelding/UbehandletMeldingUtils.java
package no.nav.sbl.dialogarena.mininnboks.provider.rest.ubehandletmelding;

import no.nav.sbl.dialogarena.mininnboks.consumer.domain.Henvendelse;

import java.util.Comparator;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;
import java.util.function.Supplier;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.toList;
import static no.nav.sbl.dialogarena.mininnboks.consumer.domain.Henvendelse.NYESTE_OVERST;
import static no.nav.sbl.dialogarena.mininnboks.provider.rest.ubehandletmelding.UbehandletMelding.*;

public class UbehandletMeldingUtils {
    static Supplier<Set<Henvendelse>> supplier = () -> new TreeSet<>(Comparator.comparing(h -> h.traadId));

    public static List<UbehandletMelding> hentUbehandledeMeldinger(List<Henvendelse> henvendelser) {

        return henvendelser
                .stream()
                .sorted(NYESTE_OVERST)
                .collect(Collectors.toCollection(supplier))
                .stream()
                .filter(erIkkeKassert)
                .filter(erUbesvart.or(erUlest))
                .map(UbehandletMelding::new)
                .collect(toList());
    }
}

File: src/main/java/no/nav/sbl/dialogarena/mininnboks/provider/rest/ubehandletmelding/UbehandletMeldingUtils.kt
package no.nav.sbl.dialogarena.mininnboks.provider.rest.ubehandletmelding

import no.nav.sbl.dialogarena.mininnboks.consumer.domain.Henvendelse
import java.util.*
import java.util.function.Supplier
import java.util.stream.Collectors

object UbehandletMeldingUtils {
    var supplier = Supplier<Set<Henvendelse>> { TreeSet(Comparator.comparing { h: Henvendelse -> h.traadId!! }) }
    fun hentUbehandledeMeldinger(henvendelser: List<Henvendelse?>): List<UbehandletMelding> {
        return henvendelser
                .stream()
                .sorted(Henvendelse.NYESTE_OVERST)
                .collect(Collectors.toCollection(supplier))
                .stream()
                .filter(UbehandletMelding.erIkkeKassert)
                .filter(UbehandletMelding.erUbesvart.or(UbehandletMelding.erUlest))
                .map { henvendelse: Henvendelse? -> UbehandletMelding(henvendelse!!) }
                .collect(Collectors.toList())
    }
}
